Proposal: working group for reference types (#103)
opencontainers/tob: proposals/wg-reference-types.md

This document describes the basic governance principles for the Reference Types Working Group (the “WG”).

The WG operates as an OCI Working Group under the [Open Container Initiative (OCI) Charter](https://github.com/opencontainers/tob/blob/master/CHARTER.md), which describes the responsibilities of the OCI Technical Oversight Board (the "TOB”). The WG is established by the TOB as an OCI Working Group pursuant to the OCI Charter. Accordingly, the WG will operate in accordance with the OCI Charter and OCI's other policies and procedures, supplemented by the details below.

## Purpose

As cloud native development continues to grow, there is increased interest in evolving registries to natively store, discover, and pull a graph of content associated with specific container images in a registry. Use cases for said associated artifacts include but are not limited to Software Bill of Materials (SBoM), security scan results, and signing. Having a native way to store and discover artifacts associated with other artifacts enables end-users to answer the question of: “What SBOMs or signatures are associated with this container image?”

## Scope

> 

Reviewer: They are supposed to be indented from the second bullet on.

* Define and deliver the capability to store, discover, and pull a graph of artifacts associated with a specific artifacts to OCI distribution compliant registries. These set of capabilities has been commonly known as "reference types" or "references".
  * Define supported use cases
  * Document impact to existing user-facing tools and registries
  * Define the method for creating, distributing, and discovering referenced objects
  * Document user expectations for promoting an artifact between registries with its references

>   * Document onboarding process for registries and user-facing tools to adopt reference types

Reviewer 2: ```suggestion
  * Document path to adoption for registries and other implementations
```

  * Defined expectations for artifact reference lifecycle management

>   * Deliver a reference implementation of the reference types proposal

Reviewer 3: Just a note per the Governance section and the relevant parts of the Charter: a new reference implementation will require a new TOB vote at the time it is ready for it to become an OCI Project.

## Out of Scope

* Identified out of scope items will be listed here as WG progresses

## Intended work product

* Referrers API specification that provides the ability to discover references to existing container images. These include listing signatures, SBoMs, security scan results that refer to the digest of a manifest. The referrers API specification will sit within or along side the Distribution specification.

> * Identify, and document the pros and cons for versioning the existing manifests, compared with creating a new manifest to support reference types.

Reviewer 2: ```suggestion
* Proposal to extend existing manifests or create a new manifest to support reference types.
```

Author: To address any of the existing manifests, we do need to address how these would be versioned. So, I propose we leave this as stated as the updated wording does change the process.

Reviewer 2: Just documenting the pros and cons doesn't feel like a great work product. Having a concrete proposal for existing manifests feels like a prerequisite for doing this analysis, so if that's implied by what you have here, that's fine with me.

Author: Ahh, I see, so it's a merger of the two. 
What I didn't want to lose was the versioning analysis, and just jump into debating solutions. I've merged both concepts.

* Identify, and document the pros and cons for versioning the existing manifests, compared with creating a new manifest to support reference types.

## Proposed Owners
* Lachlan Evenson (@lachie83)
* Justin Cormack (@justincormack)
* Michael Brown (@michaelb990)
* Derek McGowan (@dmcgowan)
* Jon Johnson (@jonjohnsonjr)

## Sponsors

* Microsoft
* AWS
* Docker

> 

Author: From #96, there's an open clarification if Google will be a sponsor.

Reviewer 3: If there's an open question we should resolve that before voting.

Author: @jonjohnsonjr, can you clarify if Google will be a sponsor?

Reviewer 4: Would it effect the vote either way? I'm no longer at Google, but if I remember the process correctly, getting an actual answer here will be very annoying and time consuming.

Reviewer 3: I'd rather not start the voting process if we expect the thing we're voting on to change.  If @SteveLasker wants to drop the question and we vote on the proposal as-is that would be fine to me.

Author: I just found it weird that we had all but one owner in sync with sponsors. 
Can @jonjonsonjr, or someone from Google make that call? It doesn't seem like that's my decision to make.

Reviewer 2: There's also an open clarification regarding the set of sponsors being exclusively corporations: https://github.com/opencontainers/tob/issues/96#issuecomment-965897973

The only mention of sponsorship in the [WG requirements](https://github.com/opencontainers/tob/blob/e67eab0f4264c4f7c02e8aa04cecd7c7349604a5/WG-INFO.md) I can find says:

> a list of OCI Projects, non-OCI projects, or organizations sponsoring the working group and participating in the implementation and use case validation of the work done by the group

I can't speak on behalf of Google as a whole, and I expect anyone who _can_ would want a much more formal definition of sponsorship. I'm happy for the vote to proceed without Google as a sponsor, but I'm not happy with the current list of sponsors. I'd expect to see sponsorship from oras, notation, cosign, harbor, distribution, quay, distribution-spec, image-spec, etc. -- not just companies.

Reviewer 4: I'm happy to add cosign to the list if that helps. I'm also happy to leave it off if that helps get the vote started faster.

Reviewer 5: If it's a point of contention, I suggest we just remove any notion of "corporations" as sponsors as the work done under this WG has been discussed at length without any specific need for guidance/sponsorship from the employers/entities themselves.

I understand initially the list was denoting companies which a specific interest in this work and had people involved for that purpose, but I think we've made it clear this is an open WG that will hopefully involve input from any registry operator, open source registry implementation, clients, etc.

Delaying at this point to formalize or determine a list of company names seems unnecessary given we have struggled for months to get this rolling and are all fully aware of the work that needs to be accomplished collaboratively no matter who is listed in this proposal. My intent was to hopefully involve a good cross-section of people, which should be much larger than just the "owners", but the owners list will help shepherd the discussion, status, and progress of the group and, in my opinion, is a good cross-section of people with OCI history, registry open source projects, operating registries, and so on.

Reviewer 6: exactly what phil said. While I'm sure that "sponsor" equates to "my employer allows me to allocate time to this, and is interested in seeing this completed", it would be tedious to tease that out before getting this in.
I'd rather just see GH handles of the stakeholders that are going to see this through to completion.

Reviewer 6: can't we just change this "Sponsor" to "Stakeholder" and move on with it?

Author: Changed sponsors to stakeholders. 
With this change, are there any other changes before moving to the vote?

## Related issues/PRs

* https://github.com/opencontainers/tob/issues/96
* https://github.com/opencontainers/artifacts/pull/27
* https://github.com/opencontainers/artifacts/pull/29
* https://github.com/opencontainers/artifacts/pull/37
* https://github.com/opencontainers/image-spec/issues/827
* https://github.com/opencontainers/image-spec/pull/828

> * https://github.com/oras-project/artifacts-spec/

Reviewer 3: Just a note here that this repository is not itself an OCI Project even though many of the people working on it are also involved in OCI (and in this proposed WG).
